capstone/src/main/java/com/google/step/search/ElasticsearchClient.java

```diff
@@ -55,21 +64,106 @@ public ElasticsearchClient(String hostname, short port) {
   }
 
   /**
-   * Given a {@code RestaurantHeader}, sends HTTP request to Elasticsearch server to add a document
-   * to the "restaurants" index representing a {@code restaurantHeader}.
-   * @param restaurantHeader a {@code RestaurantHeader}
+   * Given a {@link RestaurantHeader}, sends HTTP request to the search server to add a document
+   * that represents a {@link RestaurantHeader} to the "restaurants" index.
+   * @param restaurantHeader a {@link RestaurantHeader} that will be updated in the search index
    * @throws IOException when buildPutRequest() or execute() fails
    */
   public void updateRestaurantHeader(RestaurantHeader restaurantHeader) throws IOException {
     String restaurantKey = String.valueOf(restaurantHeader.getRestaurantKey());
+    List<String> requestPath = Arrays.asList("", RESTAURANTS, "_doc", restaurantKey);
 
+    String requestBody = gson.toJson(restaurantHeader);
+
+    HttpRequest request = buildElasticsearchHttpRequest("PUT", requestPath, requestBody);
+    request.execute();
+  }
+
+  /**
+   * Queries search server for {@link RestaurantHeader} fields that match {@code query} by either
+   * "name" or "cuisine" field.
+   * @param query valid query string
+   * @return list of {@link RestaurantHeader} objects sorted by descending relevance score
+   * @throws IOException if request cannot be made or executed properly
+   */
+  public List<RestaurantHeader> queryRestaurantHeaders(String query) throws IOException {
+    List<String> requestPath = Arrays.asList("", RESTAURANTS, "_search");
+
+    String requestBody = new JSONObject()
+        .put("query", new JSONObject()
+            .put("multi_match", new JSONObject()
+                .put("query", query)
+                .put("fields", new JSONArray(Arrays.asList("name", "cuisine")))))
+        .toString();
+
+    HttpRequest request = buildElasticsearchHttpRequest("POST", requestPath, requestBody);
+    HttpResponse response = request.execute();
+
+    return convertElasticsearchResponseBodyToHeaders(response);
+  }
+
+  /**
+   * Queries search server, matching everything possible.
+   * @return list of {@link RestaurantHeader} objects sorted arbitrarily
+   * @throws IOException if request cannot be made or executed properly
+   */
+  public List<RestaurantHeader> getRandomRestaurants() throws IOException {
```

Comment:
The solution here is more temporary than anything. I was going to put a limit on the number of results returned in a later pull request just so that too much data isn't returned.

Reply:
Also, I couldn't think of a better name for the method that allows for more flexibility in its use than specifically for the landing page. I feel like its process is general enough that calling it something along the lines of `getLandingPageRestaurantHeaders()` would not really be suitable.

Reply:
I was thinking about this, and unless Megan's experience with search provides better long term advice, you might want to have a "searchRestaurants" method that takes a type that you can later update to include more things. For example:

`public List<RestaurantHeader> searchRestaurants(String searchTerm, SearchAttributes attributes) { ... }`

And then SearchAttributes could include specific things later, like distance from a location, cuisine types, etc.

Then, calling that with an empty string would just result in getting all matching documents (i.e. all of them).

Reply:
That might help in terms of simplifying the API interface, but I think I would still end up having to implement a bunch of different functions as I am now. It's surprising how differently the request bodies are formatted for a given kind of query. Should I maybe have an overloaded function that doesn't take into account search attributes? I feel that incorporating search attributes is a great idea in terms of simplifying the design, but it'll take a bit of work to implement properly.

Reply:
Yeah, the single search function makes sense. Since I don't currently have any way to encode search attributes, should I just have a single argument for now, and add an overload with the search attributes argument once it's actually implemented?